### DISCUSSION TITLE
Docker action persistence not working as expected

### DISCUSSION THREAD

**Discussion by Derongan** [community/community#26057] [Actions]
<p>I have a docker action that is copying data into $HOME</p>
<pre><code>...
# Install to github .m2
mkdir -p $HOME/.m2/repository
echo "Copying from /root/.m2/repository to $HOME/.m2/repository"
cp -a /root/.m2/repository/. $HOME/.m2/repository
</code></pre>
<p>However when I add</p>
<pre><code>- run: ls -lah $HOME/.m2
</code></pre>
<p>to my workflow, the repository directory is not found. Why is this not being persisted, and what are the workarounds?</p>

**Comment by BrightRan** [comment]
<p><a class="mention" href="/derongan">@derongan</a> ,</p>
<p>I tested and tried the commands <strong>mkdir</strong> , <strong>ls</strong> and <strong>cp</strong> on my repository like as below:</p>
<pre><code>jobs:
  TEST:
    name: Test
    runs-on: ubuntu-latest
    steps:
      - name: Checkout
        uses: actions/checkout@v2

      - name: Create directory
        run: |
          echo "run mkdir -p $HOME/.m2/repository"
          mkdir -p $HOME/.m2/repository
          echo " ********************************ls in #HOME********************************"
          echo $HOME
          ls -lah $HOME
          echo " ******************************ls in $HOME/.m2******************************"
          ls -lah $HOME/.m2
      - name: Copy files
        run: |
          echo "run cp -a TestClock/. $HOME/.m2/repository"
          cp -a TestClock/. $HOME/.m2/repository
          echo " ******************************ls in $HOME/.m2******************************"
          ls -lah $HOME/.m2
          echo " *************************ls in $HOME/.m2/repository*************************"
          ls -lah $HOME/.m2/repository
</code></pre>
<p>and this is the result:</p>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/5/5601322b244d62f49d566b54cd8fb8246b8ee94a.png" data-download-href="/uploads/short-url/cgPCJMISZ1o3UBFSG4qRfqNWBXI.png?dl=1" title="results.png" rel="nofollow noopener"><img src="https://user-images.githubusercontent.com/46307996/181095005-4018da08-3b11-469e-9c89-189da8d73b68.png" alt="results.png" data-base62-sha1="cgPCJMISZ1o3UBFSG4qRfqNWBXI" width="690" height="264" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/5/5601322b244d62f49d566b54cd8fb8246b8ee94a_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">results.png</span><span class="informations">1174×450 44.8 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>
<p>You can see I run the <strong>ls</strong> command after using <strong>mkdir</strong> command to create the directory and after using <strong>cp</strong> command to copy files. Looks like all the commands can work fine on my side, and did not occur the problem as you reported.</p>
<p>Maybe you also can try the steps like as that in my example, to check whether the directory  <strong>$HOME/.m2/repository</strong> is really created successfully after execute the <strong>mkdir</strong> command. And also after copying the files.</p>
<p>And if possible, please also share your repository with us, so that we can check the complete configurations of your workflow and the detailed debug logs of the workflow run to analyze the root cause.</p>

**Comment by BrightRan** [comment, marked as answer]
<p><a class="mention" href="/derongan">@derongan</a> ，</p>
<p>How are things going? Is my suggestions in the previous reply helpful to your?</p>
<p>Let me add some more points here:</p>
<ol>
<li>
<p>If you execute these commands on a Docker container (or a Docker container action), when building and running the container, it will automatically mount volumes on the container to share files between the runner machine. From the logs of the workflow run, you can see the command line like as below example:</p>
<p>/usr/bin/docker run … -v “/home/runner/work/_temp/_github_home”:"/github/home" -v “/home/runner/work/_temp/_github_workflow”:"/github/workflow" -v “/home/runner/work/TestClock/TestClock”:"/github/workspace" …</p>
</li>
</ol>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/e/e9a6fbb79b47f085764d4dfc4f4824f0c443d55a.png" data-download-href="/uploads/short-url/xkZ0EPnyFWtdTUCaCwWBh27Lb2W.png?dl=1" title="example.png" rel="nofollow noopener"><img src="https://user-images.githubusercontent.com/46307996/181095006-6b02c92f-69f7-410e-940d-445fae70b935.png" alt="example.png" data-base62-sha1="xkZ0EPnyFWtdTUCaCwWBh27Lb2W" width="690" height="88" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/e/e9a6fbb79b47f085764d4dfc4f4824f0c443d55a_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">example.png</span><span class="informations">901×116 9.69 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>
<p>This means files updated will be synced between the mounted volumes. For example update files in the directory " <strong>/github/home</strong>" on the container will be synced to the directory " <strong>/home/runner/work/_temp/_github_home</strong>" on the runner machine.</p>
<ol start="2">
<li>On the <strong>container</strong> , the <strong>$HOME</strong> is the directory " <strong>/github/home</strong>", if run the mkdir command on the container ( <strong>mkdir -p $HOME/.m2/repository</strong> ), it will generate the directory " <strong>/github/home/.m2/repository</strong>",and also sync to the runner machine to generate the directory " <strong>/home/runner/work/_temp/_github_home/.m2/repository</strong>". However, on the <strong>runner machine</strong> , the <strong>$HOME</strong> is the directory " <strong>/home/runner</strong>".</li>
</ol>
<p>According to the above points, if you execute the command " <strong>mkdir -p $HOME/.m2/repository</strong>" in your Docker container action, and then execute the command " <strong>ls -lah $HOME/.m2</strong>" in another step which runs on the runner machine, not on the container, you will definitely get the “directory not found"error. You should view the file list in the directory " <strong>/home/runner/work/_temp/_github_home</strong>” on runner machine, instead of in <strong>$HOME</strong>.</p>
<p>Hope this reply can help you.</p>

**Comment by Derongan** [comment]
<p>Hello,</p>
<p>I have not had a chance to play around yet with your solution. I am still playing around with a number of things with actions, but will get to this once I figure those out. Thanks!</p>

**Comment by BrightRan** [comment]
<p><a class="mention" href="/derongan">@derongan</a> ,</p>
<p>All right, if you have any progress, feel free to tell me.</p>
<p>However, as I mentioned in my second reply, I think the reason causes the problem you are facing is that you created directory with the <strong>$HOME</strong> on the container but when viewing file list you used the $ <strong>HOME</strong> on the runner machine. Because the environment variable  <strong>$HOME</strong>  on the container and on the runner machine points to different directories.</p>

**Comment by jef** [comment]
<aside class="quote group-github-partners" data-username="brightran" data-post="3" data-topic="17745">
<div class="title">
<div class="quote-controls"></div>
 brightran:</div>
<blockquote>
<p>According to the above points, if you execute the command " <strong>mkdir -p $HOME/.m2/repository</strong> " in your Docker container action, and then execute the command " <strong>ls -lah $HOME/.m2</strong> " in another step which runs on the runner machine, not on the container, you will definitely get the “directory not found"error. You should view the file list in the directory " <strong>/home/runner/work/_temp/_github_home</strong> ” on runner machine, instead of in <strong>$HOME</strong> .</p>
</blockquote>
</aside>
<p>This is great information for test reports. Thank you for this!!!</p>
